File: core/security.py
# -*- coding: utf-8 -*-
"""
Security validation module for ZipToTxt.
Defends against Zip Slip, Zip Bomb, Symlink escaping, Windows device names,
Unicode Trojan Source / Bidi overrides, and sensitive credential exposure.
"""
import hashlib
import logging
import re
import urllib.parse
from typing import Union
from .constants import (
    SENSITIVE_DIR_NAMES,
    SENSITIVE_FILE_NAMES,
    WINDOWS_RESERVED_NAMES,
    IGNORED_METADATA_FILES,
)
from .models import SecurityError

logger = logging.getLogger(__name__)

def human_size(num_bytes: Union[int, float]) -> str:
    """Formats bytes into human readable binary units."""
    size = float(num_bytes)
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024.0 or unit == 'TB':
            return f"{int(size)} B" if unit == 'B' else f"{size:.1f} {unit}"
        size /= 1024.0
    return f"{size:.1f} PB"

def calculate_sha256(data: bytes) -> str:
    """Calculates SHA-256 hex digest for byte payload."""
    return hashlib.sha256(data).hexdigest()

def normalize_ai_path(raw_path: str) -> str:
    """
    Sanitizes, normalizes and validates file paths to prevent directory traversal (Zip Slip),
    Unicode Trojan Source, Windows reserved device names, null-byte injection, and absolute paths.
    """
    if not raw_path or not isinstance(raw_path, str):
        raise SecurityError("Path cannot be empty.")

    # 1. Clean hidden control chars, zero-width spaces, bidi overrides, null bytes
    cleaned = re.sub(r'[\x00-\x1f\x7f\u200b-\u200f\u202a-\u202e\ufeff]', '', raw_path).strip()

    # 2. Iteratively decode multi-level URL encoded sequences (e.g. %252e%252e, %2e%2e, %2f, %5c)
    decoded = cleaned
    for _ in range(3):
        if '%' in decoded:
            try:
                nxt = urllib.parse.unquote(decoded)
                if nxt == decoded:
                    break
                decoded = nxt
            except Exception:
                break
        else:
            break
    path = decoded

    # 3. Replace backslashes with forward slashes and strip markdown quotes/brackets
    path = path.replace('\\', '/')
    path = re.sub(r'^[`"\'\s*#\[\(]+|[`"\'\s*#\]\)]+$', '', path).strip()

    if not path:
        raise SecurityError(f"Invalid empty file path from raw input: {raw_path!r}")

    # Disallow UNC network shares and Windows drive letters (C:, D:, //server/...)
    if re.match(r'^[A-Za-z]:', path) or path.startswith('//'):
        raise SecurityError(f"Absolute or UNC network paths are forbidden: {raw_path}")

    # 4. Remove leading slashes and relative './' prefixes
    path = re.sub(r'^/+', '', path)
    while path.startswith('./'):
        path = re.sub(r'^/+', '', path[2:])

    if not path:
        raise SecurityError(f"Path resolved to empty relative path: {raw_path}")

    # 5. Strict Zip Slip check on components
    raw_parts = path.split('/')
    parts = []
    for part in raw_parts:
        trimmed = part.strip()
        if not trimmed or trimmed == '.':
            continue
        if trimmed == '..' or trimmed.lower() == '%2e%2e' or '..' in trimmed:
            raise SecurityError(f"Directory traversal sequence (Zip Slip) detected: {raw_path}")
        parts.append(trimmed)

    if not parts:
        raise SecurityError(f"Path resolved to empty directory root: {raw_path}")

    # 6. Check Windows reserved device names (CON, PRN, AUX, NUL, COM1-9, LPT1-9)
    for part in parts:
        base_name = part.split('.')[0].upper()
        if base_name in WINDOWS_RESERVED_NAMES:
            raise SecurityError(f"Windows reserved device name is forbidden: {part} in {raw_path}")

    return "/".join(parts)

def is_safe_relative_path(path: str) -> bool:
    """Returns True if the path is safe from traversal and reserved names."""
    try:
        normalize_ai_path(path)
        return True
    except SecurityError:
        return False

def is_sensitive_path(rel_path: str) -> bool:
    """Checks if the path contains sensitive keys, env vars, or private configuration."""
    try:
        norm = normalize_ai_path(rel_path)
        parts = norm.split('/')
        if any(p.lower() in SENSITIVE_DIR_NAMES for p in parts):
            return True
        file_name = parts[-1].lower()
        if file_name in SENSITIVE_FILE_NAMES:
            return True
        if (
            file_name.startswith('.env') or
            file_name.endswith(('.pem', '.key', '.pfx', '.p12')) or
            'id_rsa' in file_name or
            'id_ed25519' in file_name or
            'service_account' in file_name or
            'service-account' in file_name
        ):
            return True
        return False
    except Exception:
        return True

def is_ignored_metadata_path(rel_path: str) -> bool:
    """Checks if the path is OS metadata (e.g. __MACOSX, .DS_Store, Thumbs.db)."""
    parts = rel_path.split('/')
    for p in parts:
        if p in IGNORED_METADATA_FILES or p.startswith('._') or p == '__MACOSX':
            return True
    return False

File: src/utils/security.ts
import {
  SENSITIVE_DIR_NAMES,
  SENSITIVE_FILE_NAMES,
  WINDOWS_RESERVED_NAMES,
  IGNORED_METADATA_FILES,
} from './constants';

export function humanSize(value: number): string {
  let size = Number(value);
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  for (let i = 0; i < units.length; i++) {
    const unit = units[i];
    if (size < 1024 || i === units.length - 1) {
      return unit === 'B' ? `${Math.floor(size)} B` : `${size.toFixed(1)} ${unit}`;
    }
    size /= 1024;
  }
  return `${size.toFixed(1)} PB`;
}

/**
 * Sanitizes and normalizes file paths to prevent directory traversal (Zip Slip),
 * Unicode Trojan Source attacks, Windows reserved device names, and null-byte injection.
 */
export function normalizeAiPath(rawPath: string): string {
  if (!rawPath || typeof rawPath !== 'string') {
    throw new Error('文件路径不能为空');
  }

  // 1. Remove dangerous zero-width, bidi override, control and null characters
  let path = rawPath
    .replace(/[\u0000-\u001F\u007F\u200B-\u200F\u202A-\u202E\uFEFF]/g, '')
    .trim();

  // 2. Iteratively decode multi-level URL encoded traversal attempts (e.g. %252e%252e, %2e%2e, %2f, %5c)
  let decoded = path;
  for (let i = 0; i < 3; i++) {
    if (/%[0-9a-fA-F]{2}/.test(decoded)) {
      try {
        const next = decodeURIComponent(decoded);
        if (next === decoded) break;
        decoded = next;
      } catch {
        break;
      }
    } else {
      break;
    }
  }
  path = decoded;

  // 3. Normalize backslashes to forward slashes & trim markdown code markers / quotes / brackets
  path = path
    .replace(/\\/g, '/')
    .replace(/^[`"'\s*#\[\(]+|[`"'\s*#\]\)]+$/g, '')
    .trim();

  if (!path) {
    throw new Error(`非法空文件路径: ${JSON.stringify(rawPath)}`);
  }

  // Disallow UNC network shares and drive letters (C:, D:, //server/...)
  if (/^[A-Za-z]:/.test(path) || path.startsWith('//')) {
    throw new Error(`必须使用仓库相对路径，禁止绝对路径与网络共享路径: ${rawPath}`);
  }

  // 4. Remove leading slashes and relative './' or '.\\' prefixes
  path = path.replace(/^\/+/, '');
  while (path.startsWith('./')) {
    path = path.slice(2).replace(/^\/+/, '');
  }

  if (!path) {
    throw new Error(`非法空文件路径: ${JSON.stringify(rawPath)}`);
  }

  // 5. Strict path traversal parts (Zip Slip defense)
  const rawParts = path.split('/');
  const parts: string[] = [];

  for (const part of rawParts) {
    const trimmedPart = part.trim();
    if (!trimmedPart || trimmedPart === '.') {
      continue;
    }
    // Prevent Zip Slip variations like "..", "...", "%2e%2e", or parts containing dangerous traversal sequences
    if (
      trimmedPart === '..' ||
      trimmedPart.toLowerCase() === '%2e%2e' ||
      trimmedPart.includes('..')
    ) {
      throw new Error(`检测到路径穿越攻击 (Zip Slip) 被严格拦截: ${rawPath}`);
    }
    parts.push(trimmedPart);
  }

  if (parts.length === 0) {
    throw new Error(`路径解析后为空，非法路径: ${rawPath}`);
  }

  // 6. Check Windows reserved device names (CON, PRN, AUX, NUL, COM1..9, LPT1..9)
  for (const part of parts) {
    const baseName = part.split('.')[0].toUpperCase();
    if (WINDOWS_RESERVED_NAMES.has(baseName)) {
      throw new Error(`禁止使用系统保留设备名: ${part} (${rawPath})`);
    }
  }

  return parts.join('/');
}

/**
 * Checks if a path points to sensitive credentials, git repositories, or key files.
 */
export function isSensitivePath(relPath: string): boolean {
  try {
    const normalized = normalizeAiPath(relPath);
    const parts = normalized.split('/');
    
    // Check directory names
    if (parts.some(p => SENSITIVE_DIR_NAMES.has(p.toLowerCase()))) {
      return true;
    }
    
    const fileName = parts[parts.length - 1].toLowerCase();
    
    // Check sensitive file names or patterns
    if (SENSITIVE_FILE_NAMES.has(fileName)) {
      return true;
    }

    if (
      fileName.startsWith('.env') ||
      fileName.endsWith('.pem') ||
      fileName.endsWith('.key') ||
      fileName.endsWith('.pfx') ||
      fileName.endsWith('.p12') ||
      fileName.includes('id_rsa') ||
      fileName.includes('id_ed25519') ||
      fileName.includes('service_account') ||
      fileName.includes('service-account')
    ) {
      return true;
    }

    return false;
  } catch {
    return true;
  }
}

/**
 * Checks whether a file path is junk OS metadata that should be ignored by default
 */
export function isIgnoredMetadataPath(relPath: string): boolean {
  const parts = relPath.split('/');
  for (const p of parts) {
    if (IGNORED_METADATA_FILES.has(p) || p.startsWith('._') || p === '__MACOSX') {
      return true;
    }
  }
  return false;
}

export async function calculateSha256(buffer: ArrayBuffer): Promise<string> {
  const hashBuffer = await crypto.subtle.digest('SHA-256', buffer);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
}

export function arrayBufferToBase64(buffer: ArrayBuffer): string {
  let binary = '';
  const bytes = new Uint8Array(buffer);
  const len = bytes.byteLength;
  for (let i = 0; i < len; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}
